File: day2/2.hs
import Data.Char (isDigit, isAlpha)
import Data.List (groupBy, elemIndex, sortBy)
import Data.Maybe (fromJust, mapMaybe)
import Data.Function (on)

readLines :: FilePath -> IO [String]
readLines file = do
  contents <- readFile file
  return $ lines contents

numbers :: String -> [Int]
numbers = map read . filter (all isDigit) . groupBy (\x y -> isDigit x == isDigit y)

alwords :: String -> [String]
alwords = filter (all isAlpha) . groupBy (\x y -> isAlpha x == isAlpha y)

colours :: String -> [String] -> [Int]
colours line options = mapMaybe (`elemIndex` options) (alwords line)

gameMax :: String -> [String] -> [Int]
gameMax g c = head x : map (maximum . map snd) (groupBy ((==) `on` fst) (sortBy (compare `on` fst) y))
  where x = numbers g; y = zip (colours g c) (tail x)

p1 :: String -> Maybe Int
p1 line = if and $ zipWith (<=) xs [12, 13, 14] then Just x else Nothing
  where x : xs = gameMax line ["red", "green", "blue"]

p2 :: String -> Int
p2 line = product (tail (gameMax line ["red", "green", "blue"]))

main :: IO ()
main = do
  lines <- readLines "input.txt"
  print $ sum $ mapMaybe p1 lines
  print $ sum $ map p2 lines

File: day2/2.c

#include "../a.h"

SV gm(I gmx[4], C * b) {
    W(!isdigit(*b))b++;gmx[0]=atoi(b);W(isdigit(*b))b++;
    W(*b) {
        W(!isdigit(*b))b++;I num=atoi(b);W(isdigit(*b))b++;b++;
        if(!strncmp(b,"red",3)){IF(num>gmx[1],gmx[1]=num);b+=3;}
        else if(!strncmp(b,"green",5)){IF(num>gmx[2],gmx[2]=num);b+=5;}
        else if(!strncmp(b,"blue",4)){IF(num>gmx[3],gmx[3]=num);b+=4;}}}

SI p1(I gmx[4])_(gmx[1]>12||gmx[2]>13||gmx[3]>14?0:gmx[0])
SI p2(I gmx[4])_(gmx[1]*gmx[2]*gmx[3])

I main(V) {
    FILE*input=fopen("input.txt","r");
    I sum1=0,sum2=0;
    W (!feof(input)) {
        C buf[256]={0};fgets(buf,256,input);IF(!*buf,continue)
        I gmx[4]={0};gm(gmx,buf);sum1+=p1(gmx),sum2+=p2(gmx); }
    printf("%d\n%d\n",sum1,sum2); }

